task/move.go

```diff
@@ -86,27 +87,82 @@ func Move(fetch Fetcher, put Putter) {
 	}
 }
 
+func (o *oldStore) fetchB() (thing Thing, ok bool, err error) {
+	// simulate the delay
+	time.Sleep(3 * time.Second)
+	fmt.Println("fetchB: fetching...", o.bookNo)
+
+	if o.bookNo == 5 {
+		return thing, ok, errors.New("Fetch encountered network issues")
+	}
+
+	if o.bookNo > len(o.bookInventory)-1 {
+
+		return thing, ok, err
+	}
+
+	thing, ok = o.bookInventory[o.bookNo], true
+	o.bookNo++
+	return
+}
+
+func (n *newStore) putB(thing Thing) error {
+	time.Sleep(3 * time.Second)
+
+	n.inventory = append(n.inventory, thing)
+
+	if len(n.inventory) == 5 {
+		return errors.New("store is filled up")
+	}
+
+	return nil
+}
+
 // MaybeMove is exactly the same as Move except that it may return an error
 // because fetch() and put() may return errors. If no errors occur then
 // MaybeMove returns under the same conditions as Move(). If an error occurs
 // then MaybeMove returns earlier even if there are more Things to fetch().
 func MaybeMove(fetch MaybeFetcher, put MaybePutter) error {
-	// [TODO]
+	ch := make(chan Thing)
+	errCh := make(chan error, 2)
+	quit := make(chan struct{})
```

Comment:
If there is never an error in the `put()` loop then `quit` is never closed so you leak a resource. The problem is that you can't double-close a channel because it will panic. You can, however, double cancel a context:

```Go
ctx, cancel := context.WithCancel(context.Background())
defer cancel()
…
// fetch() loop
select {
  case <-ctx.Done():
    return
}

if err := put(thing); err != nil {
  cancel()
  …
}
```

Reply:
This would also allow you to accept a `context.Context` like in `MoveCtx()` and use it instead of `context.Background()` so that the caller can cancel it themselves.